// firebase-tools/lib/mcp/tools/core/get_environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.get_environment = void 0;
const zod_1 = require("zod");
const tool_1 = require("../../tool");
const util_1 = require("../../util");
const projectUtils_1 = require("../../../projectUtils");
const js_yaml_1 = require("js-yaml");
const auth_1 = require("../../../auth");
const configstore_1 = require("../../../configstore");
exports.get_environment = (0, tool_1.tool)({
    name: "get_environment",
    description: "Use this to retrieve the current Firebase **environment** configuration for the Firebase CLI and Firebase MCP server, including current authenticated user, project directory, active Firebase Project, and more.",
    inputSchema: zod_1.z.object({}),
    annotations: {
        title: "Get Firebase Environment Info",
        readOnlyHint: true,
    },
    _meta: {
        requiresAuth: false,
        requiresProject: false,
    },
}, async (_, { projectId, host, accountEmail, rc, config }) => {
    const aliases = projectId ? (0, projectUtils_1.getAliases)({ rc }, projectId) : [];
    const geminiTosAccepted = !!configstore_1.configstore.get("gemini");
    return (0, util_1.toContent)(`# Environment Information

Project Directory: ${host.cachedProjectDir}
Project Config Path: ${config.projectFileExists("firebase.json") ? config.path("firebase.json") : "<NO CONFIG PRESENT>"}
Active Project ID: ${projectId ? `${projectId}${aliases.length ? ` (alias: ${aliases.join(",")})` : ""}` : "<NONE>"}
Authenticated User: ${accountEmail || "<NONE>"}
Gemini in Firebase Terms of Service: ${geminiTosAccepted ? "Accepted" : "Not Accepted"}

# Available Project Aliases (format: '[alias]: [projectId]')

${(0, js_yaml_1.dump)(rc.projects).trim()}

# Available Accounts:

${(0, js_yaml_1.dump)((0, auth_1.getAllAccounts)().map((account) => account.user.email)).trim()}
${config.projectFileExists("firebase.json")
        ? `
# firebase.json contents:

\`\`\`json
${config.readProjectFile("firebase.json")}
\`\`\``
        : `\n# Empty Environment

It looks like the current directory is not initialized as a Firebase project. The user will most likely want to:

a) Change the project directory using the 'firebase_update_environment' tool to select a directory with a 'firebase.json' file in it, or
b) Initialize a new Firebase project directory using the 'firebase_init' tool.`}`);
});
